src/ru/nsu/ccfit/zuev/osu/MainScene.java

```diff
@@ -900,20 +900,15 @@ public void onModifierFinished(IModifier<IEntity> pModifier, final IEntity pItem
                 var beatmap = parser.parse(false);
 
                 if (beatmap != null) {
-                    timingPoints = new LinkedList<>();
-
-                    for (final String s : beatmap.rawTimingPoints) {
-                        final TimingPoint tp = new TimingPoint(s.split(","), currentTimingPoint);
-                        timingPoints.add(tp);
-                        if (!tp.wasInderited() || currentTimingPoint == null) {
-                            currentTimingPoint = tp;
-                        }
-                    }
+                    timingControlPoints = new LinkedList<>(beatmap.controlPoints.timing.getControlPoints());
+                    effectControlPoints = new LinkedList<>(beatmap.controlPoints.effect.getControlPoints());
```

Comment:
Is there a reason to clone the control points lists here?
I'm asking because you haven't done this in `GameScene`.

Reply:
Ah I was thinking that the handling will be different here due to the parallax effect. The one in `GameScene` simply gets the currently active control point per update thread tick, which I guess can also be done here with a bit more work.

Reply:
After working on https://github.com/osudroid/osu-droid/pull/384/commits/f7959825f7fd680e80aec0524d6ff68ef889b7b3, it was apparent that the clone to a `LinkedList` is done to offset the logo beat and kiai effect. So unfortunately, this has to stay.